File: solutions/146.lru-cache.java
class LRUCache {
    public LRUCache(int capacity) {
        cap = capacity;
        map = new HashMap<>();
        ll = new Node(-1, -1);
        rr = new Node(-1, -1);
        ll.back = rr;
        rr.front = ll;
    }
    
    public int get(int key) {
        if(!map.containsKey(key)) return -1;
        var node = map.get(key);
        remove(node);
        add(node);
        return node.val;
    }
    
    public void put(int key, int value) {
        if(map.containsKey(key)) {
            var node = map.get(key);
            remove(node);
        }
        Node node = new Node(key, value);
        add(node);
        map.put(key, node);
        if(map.size() > cap) {
            var header = ll.back;
            map.remove(header.key);
            remove(header);
        }
    }
    
    class Node{
        Node front, back;
        int key, val;
        public Node(int key, int val){
            this.key = key;
            this.val = val;
            front = null;
            back = null;
        }
    }
    
    void add(Node node) { // add to backend
        var front = rr.front;
        rr.front = node;
        node.back = rr;
        node.front = front;
        front.back = node;
    }
    
    void remove(Node node){
        var back = node.back;
        var front = node.front;
        if(back==null || front==null) return;
        back.front = front;
        front.back = back;
        node.front = node.back = null;
    }

    final int cap;
    Map<Integer, Node> map;
    Node ll, rr;  // left/right boundary
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: solutions/146.lru-cache.rb
class LRUCache
  # double ended circluar queue
  class Node
    attr_accessor :key, :value, :pre, :nex

    def initialize(key, value, pre = nil, nex = nil)
      @key = key
      @value = value
      @pre = pre
      @nex = nex
    end
  end

  #   :type capacity: Integer
  def initialize(capacity)
    @capacity = capacity
    @key_to_node = {}
    @dummy = Node.new(-1, -1)
    @dummy.pre = @dummy.nex = @dummy
  end

  #   :type key: Integer
  #   :rtype: Integer
  def get(key)
    node = @key_to_node.dig(key)
    return -1 if node.nil?

    delete(node)
    insert(node)
    node.value
  end

  #   :type key: Integer
  #   :type value: Integer
  #   :rtype: Void
  def put(key, value)
    if @key_to_node.key?(key)
      node = @key_to_node[key]
      node.value = value
      delete(node)
      insert(node)
    else
      node = Node.new(key, value)
      @key_to_node[key] = node
      insert(node)
    end

    return unless @key_to_node.size > @capacity

    last_node = @dummy.pre
    delete(last_node)
    @key_to_node.delete(last_node.key)
  end

  private

  def insert(node)
    head_nex = @dummy.nex
    node.nex = head_nex
    head_nex.pre = node
    @dummy.nex = node
    node.pre = @dummy
  end

  def delete(node)
    node_pre = node.pre
    node_nex = node.nex
    node_pre.nex = node_nex
    node_nex.pre = node_pre
    node.pre = node.nex = nil
  end
end

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache.new(capacity)
# param_1 = obj.get(key)
# obj.put(key, value)
